[SILGen] Correct generic signature/substitutions of property wrapper generators within a closure.

hborla opened the pull request:
Closures are emitted with a generic signature if they capture generic parameters. Property wrappers declared within a closure need to take this into account when computing the generic signature for the property wrapper generator, and substitutions when calling the generator. This change does two things:
* Don't use the generic signature of context for the property wrapper generator if the context is a concrete closure.
* Substitute the archetypes from context when calling the property wrapper generator within a closure via `AppliedPropertyWrapperExpr` (which is used for parameter wrappers).

Previously, calling a property wrapper generator within a concrete closure in a generic context, e.g.

```swift
@propertyWrapper
struct Wrapper<Value> {
  var wrappedValue: Value
}

protocol P {}

func genericContext<T>(_: T) where T: P {
  let _: (Int) -> Void = { arg in
    @Wrapper var value: Int
    value = arg
  }
}
```

would hit an assertion failure in SILGen, or crash in IRGen if the type parameters are constrained.

Resolves: rdar://79587533

hborla committed:
[SILGen] Remove duplicated code between computing the lowered
interface type for the two different kinds of property wrapper
generator functions.

hborla committed:
[SILGen] For property wrapper generators, don't use the generic signature of
the enclosing context if the property wrapper is declared in a closure that
does not capture any generic parameters.

In this case, the enclosing closure won't have a generic signature, nor any
substitutions to forward when calling the property wrapper generator, which
previously caused an assertion failure when emitting the call because there
were no substitutions to use.

slavapestov commented on lib/SIL/IR/TypeLowering.cpp:
```diff
@@ -2447,38 +2447,27 @@ static CanAnyFunctionType getStoredPropertyInitializerInterfaceType(
+    inputType = VD->getPropertyWrapperInitValueInterfaceType()->getCanonicalType();
+  } else {
+    Type interfaceType = VD->getPropertyWrapperProjectionVar()->getInterfaceType();
+    if (interfaceType->hasArchetype())
```
If VarDecl::getInterfaceType() returns something with an archetype it indicates a bug elsewhere, since most callers don't expect this to occur. Were you seeing this fail with closure parameters in generic contexts or something weird like that?

hborla replied:
This code was there before (I just combined two functions into one because everything after computing `inputType` is repeated). I don't remember why I added it and it might just be unnecessary - I'm running the tests locally now without it to see if anything breaks.

slavapestov commented on lib/SIL/IR/TypeLowering.cpp:
```diff
@@ -2447,38 +2447,27 @@ static CanAnyFunctionType getStoredPropertyInitializerInterfaceType(
-  auto sig = DC->getGenericSignatureOfContext();
+  GenericSignature sig;
+  auto *closure = dyn_cast<AbstractClosureExpr>(DC);
+  if (!closure || closure->getCaptureInfo().hasGenericParamCaptures())
```
You probably want to use TC.getLoweredLocalCaptures() and not closure->getCaptureInfo(); the former is the transitive closure of the latter.

I believe something like the following will demonstrate the problem:

```
func outer<T>(_: T) {
  func inner() { print(T.self) }

  let closure = { inner() }
  // closure references inner(), which requires the generic context
}
```

hborla committed:
[SILGen] Account for transitive captures of generic parameters when computing
the generic signature for local property wrapper and default argument
generators.

hborla commented:
@swift-ci please smoke test

hborla commented:
@swift-ci please test source compatibility

slavapestov commented on test/SILGen/default_arguments_local.swift:
```diff
@@ -70,3 +70,16 @@ class ArtClass<T> {
 }
+
+// CHECK-LABEL: sil hidden [ossa] @$s23default_arguments_local5outeryyxlF : $@convention(thin) <T> (@in_guaranteed T) -> ()
+func outer<T>(_: T) {
```
Thanks for fixing this!

hborla commented:
@swift-ci please test Windows platform

hborla commented:
Source compatibility failures are unrelated; most of them are
```
error: type 'StaticString' does not conform to protocol 'Collection'
```
and the others are SDK issues.

hborla commented:
@swift-ci please test Windows platform
